File: arraysandbinarysearch/bs.py
import random
import math

arr = []

# create 100 rando numbers
for x in range(1, 1000):
    arr.append( math.floor(random.random()*1000))

# sort 
arr.sort()

# choose a random target to search for 
target=arr[math.floor(random.random()*1000)]

# initialize binary search
left = 0
right = len(arr) -1
middle = math.floor((left + right)/2)
print(middle)
iterations = 0

# search until we converge
while left <= right:
    iterations = iterations + 1
    middle = math.floor((left + right)/2)

    if arr[middle] == target:
        break
    
    if arr[middle] < target:
        left = middle + 1

    elif arr[middle] > target:
        right = middle - 1
    

# show num iterations and values for the middle index and the target
print(iterations)
print(str(arr[middle]) + " " + str(target))

File: arraysandbinarysearch/bs.js
let someArray = [];
for(var i = 0; i < 10; i++){
    someArray.push(i*Math.random());
}
console.log(someArray);
/// get a random element
let randomIndex = Math.floor((someArray.length-1)*Math.random());
console.log("random index is " + randomIndex);

let target=  someArray[randomIndex];
console.log("target is",target);
someArray.sort(function(a, b) {
    return a - b;
  });

console.log(someArray);

let left = 0;
let right = someArray.length-1;
let m = Math.floor((right + left)/2);
let iterations = 0;
while(someArray[m] !== target){
    m = Math.floor((right + left)/2);
    if(someArray[m] == target){
        break;
    }
    if(target < someArray[m]){
        right = m-1;
        
    }else{
        left = m+1;
    }
    
    console.log("ends are " + left + " " + right + " " + m);
    iterations++;
}
console.log("found " + target + " as " + someArray[m]  + " after " + iterations);



